### PR TITLE
State tables for Source Generators

### PR DESCRIPTION
**Pull request by chsienki**
Adds the state table data structures and tests, as well as stubbing out the valuesource[extensions] that will manipulate the tables via nodes.

### CODE REVIEW

File: src/Compilers/Core/Portable/SourceGeneration/IncrementalValueSource.cs [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    /// <summary>
+    /// Represents a value that can have transforms applied to it to construct an execution pipeline
+    /// </summary>
+    /// <remarks>
+    /// This is an opaque type that cannot be used directly. Instead an <see cref="IIncrementalGenerator" />
+    /// will receive a set of value sources when constructing it's execution pipeline. A set of extension methods
+    /// are then used to create transforms over the data that creates the actual pipeline.
+    /// </remarks>
+    /// <typeparam name="T">The type of value that this source provides access to</typeparam>
+    public struct IncrementalValueSource<T>
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/SourceGeneration/IncrementalValueSource.cs:20]:
```suggestion
    public readonly struct IncrementalValueSource<T>
```

File: src/Compilers/Core/Portable/SourceGeneration/UserFunction.cs [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    /// A Func based delegate that represents a user provided function
+    /// that has been wrapped to convert exceptions to UserCodeExceptions
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    /// <typeparam name="TResult"></typeparam>
+    /// <param name="arg"></param>
+    /// <returns></returns>
+    internal delegate TResult UserFunc<in T, out TResult>(T arg);
+
+    internal class UserFunctionException : Exception
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/SourceGeneration/UserFunction.cs:21]:
```suggestion
    internal sealed class UserFunctionException : Exception
```

File: src/Compilers/Core/Portable/SourceGeneration/UserFunction.cs [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+{
+    /// <summary>
+    /// A Func based delegate that represents a user provided function
+    /// that has been wrapped to convert exceptions to UserCodeExceptions
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    /// <typeparam name="TResult"></typeparam>
+    /// <param name="arg"></param>
+    /// <returns></returns>
+    internal delegate TResult UserFunc<in T, out TResult>(T arg);
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/SourceGeneration/UserFunction.cs:19]: Why can't we use `Func<T, TResult>` here instead of declaring a new `delegate` type?

File: src/Compilers/Core/Portable/SourceGeneration/UserFunction.cs [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    {
+        public UserFunctionException(Exception innerException)
+            : base("User provided code threw an exception", innerException)
+        {
+        }
+    }
+
+    internal static class UserFunctionExtensions
+    {
+        internal static UserFunc<TInput, TOutput> WrapUserFunction<TInput, TOutput>(this Func<TInput, TOutput> userFunction)
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/SourceGeneration/UserFunction.cs:31]: I can see how this would give us a bit of safety in making sure that we don't re-wrap a `Func` but not sure it's worth the extra type.

File: src/Compilers/Core/Portable/SourceGeneration/Nodes/ValueSourceExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Microsoft.CodeAnalysis.SourceGeneration.Nodes
+{
+    // PROTOTYPE(source-generators): these are just internal stubs for now, that let us build the API surface up
+    static class IncrementalValueSourceExtensions
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/SourceGeneration/Nodes/ValueSourceExtensions.cs:12]:
```suggestion
    internal static class IncrementalValueSourceExtensions
```

File: src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+            return new NodeStateTable<T>(compacted.ToImmutableAndFree(), isCompacted: true, _exception);
+        }
+
+        public static NodeStateTable<T> FromFaultedTable<U>(NodeStateTable<U> table)
+        {
+            Debug.Assert(table.IsFaulted);
+            return new NodeStateTable<T>(Empty._states, isCompacted: true, table._exception);
+        }
+
+        public class Builder
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs:107]:
```suggestion
        public sealed class Builder
```

File: src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    internal sealed class NodeStateTable<T> : IStateTable
+    {
+        internal static NodeStateTable<T> Empty { get; } = new NodeStateTable<T>(ImmutableArray<ImmutableArray<(T, EntryState)>>.Empty, isCompacted: true, exception: null);
+
+        //PROTOTYPE: there is no need to store the state per item
+        //           we can instead store one state per input, with
+        //           an optional set for modified states.
+        readonly ImmutableArray<ImmutableArray<(T item, EntryState state)>> _states;
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs:63]:
```suggestion
        private readonly ImmutableArray<ImmutableArray<(T item, EntryState state)>> _states;
```

File: src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+
+        private readonly bool _isCompacted;
+
+        readonly Exception? _exception;
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs:67]:
```suggestion
        private readonly Exception? _exception;
```

File: src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+            private Exception? _exception = null;
+
+            public void AddEntries(ImmutableArray<T> values, EntryState state)
+            {
+                _states.Add(values.SelectAsArray(v => (v, state)));
+            }
+
+            public void AddEntriesFromPreviousTable(NodeStateTable<T> previousTable, EntryState newState)
+            {
+                Debug.Assert(previousTable._states.Length > _states.Count);
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs:120]: Why is it okay assume that previous tables will be larger than the current one?

File: src/Compilers/Core/Portable/SourceGeneration/GeneratorAdaptor.cs [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+            }
+
+            if (oldContext.InfoBuilder.PostInitCallback is object)
+            {
+                context.RegisterForPostInitialization(oldContext.InfoBuilder.PostInitCallback);
+            }
+
+            context.RegisterExecutionPipeline((ctx) =>
+            {
+
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/SourceGeneration/GeneratorAdaptor.cs:50]: Is this deliberately meant to be empty or are you missing a PROTOTYPE comment? If deliberate consider adding a comment stating that to make it clearr.

File: src/Compilers/CSharp/Test/Semantic/SourceGeneration/StateTableTests.cs [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+            var previousTable = builder.ToImmutableAndFree();
+
+            builder = new NodeStateTable<int>.Builder();
+            builder.AddEntries(ImmutableArray.Create(10, 11), EntryState.Added);
+            builder.AddEntriesFromPreviousTable(previousTable, EntryState.Cached); // ((2, EntryState.Cached), (3, EntryState.Cached))
+            builder.AddEntries(ImmutableArray.Create(20, 21, 22), EntryState.Modified);
+            builder.AddEntriesFromPreviousTable(previousTable, EntryState.Removed); //((6, EntryState.Removed))); 
+            var newTable = builder.ToImmutableAndFree();
+
+
```
Comment (jaredpar) [inline src/Compilers/CSharp/Test/Semantic/SourceGeneration/StateTableTests.cs:71]:
```suggestion

```

File: src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+
+    /// <summary>
+    /// A data structure that tracks the inputs and output of an execution node
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    internal sealed class NodeStateTable<T> : IStateTable
+    {
+        internal static NodeStateTable<T> Empty { get; } = new NodeStateTable<T>(ImmutableArray<ImmutableArray<(T, EntryState)>>.Empty, isCompacted: true, exception: null);
+
+        //PROTOTYPE: there is no need to store the state per item
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs:60]: What, if any, invariants exist for the `_states` member below? For instance is it okay to have the same `T` appear multiple times in the same slot, once in defferent slots, etc ...

File: src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+        {
+            if (_isCompacted)
+                return this;
+
+            var compacted = ArrayBuilder<ImmutableArray<(T, EntryState)>>.GetInstance();
+            foreach (var entry in _states)
+            {
+                // we have to keep empty entries
+                // we only remove all entries at once, so only need to check the first item
+                if (entry.Length == 0 || entry[0].state != EntryState.Removed)
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs:93]: What does it mean to have `EntryState.Removed` in anything but the first index?

File: src/Compilers/Core/Portable/SourceGeneration/Nodes/DriverStateTable.cs [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+        private readonly ImmutableDictionary<object, IStateTable> _tables;
+
+        internal static DriverStateTable Empty { get; } = new DriverStateTable(ImmutableDictionary<object, IStateTable>.Empty);
+
+        private DriverStateTable(ImmutableDictionary<object, IStateTable> tables)
+        {
+            _tables = tables;
+        }
+
+        public class Builder
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/SourceGeneration/Nodes/DriverStateTable.cs:30]:
```suggestion
        public sealed class Builder
```

File: src/Compilers/Core/Portable/SourceGeneration/Nodes/DriverStateTable.cs [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+            private readonly ImmutableDictionary<object, IStateTable>.Builder _tableBuilder = ImmutableDictionary<object, IStateTable>.Empty.ToBuilder();
+
+            private readonly DriverStateTable _previousTable;
+
+            public Builder(DriverStateTable previousTable)
+            {
+                _previousTable = previousTable;
+            }
+
+            public NodeStateTable<T> GetLatestStateTableForNode<T>(IIncrementalGeneratorNode<T> source)
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/SourceGeneration/Nodes/DriverStateTable.cs:41]: What is the relationship between `IIncrementalGeneratorNode<T>` and an `IINcrementalGenerator<T>`?

File: src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+    be cached. This process is called 'compaction' and results in the actual tables which are stored
+    between runs, as opposed to the 'live' tables that exist during an update.
+
+    Modified entries are similar to added inputs, but with a subtle difference. When an input is Added
+    all outputs are unconditionally added too. However when an input is modified, the outputs may still
+    be the same (for instance something changed elsewhere in a file that had no bearing on the produced
+    output). In this case, the state table checks the results against the previously produced values,
+    and any that are found to be the same instead get a cached state, meaning no new downstream work 
+    will be produced for them. Thus a modified input is the only slot that can have differing output 
+    states.
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs:42]: This can be a bit hard to follow because it's fairly abstract. Can you try adding a concrete example at the bottom and walk through that? For example maybe talk about how a source file is processed by multiple generators (imagine two generators keyyng off diff attributes in the file) and how that would go through the table. 

Ideally though an example that would include a remove operation. That is the part I'm not quite following here.

File: src/Compilers/Core/Portable/SourceGeneration/UserFunction.cs (hunk 4 above)
Comment (chsienki) [reply src/Compilers/Core/Portable/SourceGeneration/UserFunction.cs:31]: Ok. It was just designed for internal safety to make sure we don't accidentally pass 'user' functions into code that doesn't handle exceptions properly. Happy to remove the delegate type.

File: src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs (hunk 9 above)
Comment (chsienki) [reply src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs:120]: It turns out that it isn't. This was made under the assumption that we always populate the table from the upstream, so there is always going to be data in the previous table. However this assumption breaks down if a new node is added to the graph later on (i.e. the IDE adds a new generator), so we can longer take it as an invariant.

File: src/Compilers/Core/Portable/SourceGeneration/Nodes/DriverStateTable.cs (hunk 15 above)
Comment (chsienki) [reply src/Compilers/Core/Portable/SourceGeneration/Nodes/DriverStateTable.cs:41]: `IIncrementalGenerator` is the public API the user implements. Internally we create a graph of `IIncrementalGeneratorNode<T>`s for each call to ValueSourceExtensions.* in the pipeline callback

File: src/Compilers/Core/Portable/SourceGeneration/GeneratorAdaptor.cs [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+using System.Collections.Generic;
+using System.Text;
+
+namespace Microsoft.CodeAnalysis.SourceGeneration
+{
+    /// <summary>
+    /// Adapts an ISourceGenerator to an incremental generator that
+    /// by providng an execution environment that matches the old one
+    /// </summary>
+    internal class SourceGeneratorAdaptor : IIncrementalGenerator
```
Comment (cston) [inline src/Compilers/Core/Portable/SourceGeneration/GeneratorAdaptor.cs:15]: >class [](http://example.com/codeflow?start=13&length=5)

`sealed`

File: src/Compilers/Core/Portable/SourceGeneration/GeneratorAdaptor.cs [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+            try
+            {
+                SourceGenerator.Initialize(oldContext);
+            }
+            catch (Exception)
+            {
+                // PROTOTYPE(source-generators):
+                // wrap in a user func exception?
+                throw;
+            }
```
Comment (cston) [inline src/Compilers/Core/Portable/SourceGeneration/GeneratorAdaptor.cs:41]: Why catch exceptions and re-throw from here rather than allowing the original exception to propagate?

File: src/Compilers/Core/Portable/SourceGeneration/IncrementalValueSource.cs [hunk 19]
```diff
@@ -0,0 +1,29 @@
...
+    {
+        internal readonly IIncrementalGeneratorNode<T> node;
```
Comment (cston) [inline src/Compilers/Core/Portable/SourceGeneration/IncrementalValueSource.cs:22]: >node [](http://example.com/codeflow?start=55&length=4)

Perhaps `Node` since the field is `internal`.

File: src/Compilers/Core/Portable/SourceGeneration/Nodes/DriverStateTable.cs [hunk 20]
```diff
@@ -0,0 +1,73 @@
...
+        internal static DriverStateTable Empty { get; } = new DriverStateTable(ImmutableDictionary<object, IStateTable>.Empty);
+
+        private DriverStateTable(ImmutableDictionary<object, IStateTable> tables)
+        {
+            _tables = tables;
+        }
+
+        public class Builder
+        {
+            private readonly ImmutableDictionary<object, IStateTable>.Builder _tableBuilder = ImmutableDictionary<object, IStateTable>.Empty.ToBuilder();
```
Comment (cston) [inline src/Compilers/Core/Portable/SourceGeneration/Nodes/DriverStateTable.cs:32]: >ImmutableDictionary<object, IStateTable>.Empty.ToBuilder [](http://example.com/codeflow?start=94&length=56)

Minor: Perhaps `ImmutableDictionary.CreateBuilder<object, IStateTable>()` instead.

File: src/Compilers/Core/Portable/SourceGeneration/Nodes/IIncrementalGeneratorNode.cs [hunk 21]
```diff
@@ -0,0 +1,19 @@
...
+
+using System.Collections.Generic;
+
+namespace Microsoft.CodeAnalysis
+{
+    /// <summary>
+    /// Represents a step in the execution piepline of an incremental generator
+    /// </summary>
+    /// <typeparam name="T">The type of value this step operates on</typeparam>
+    internal interface IIncrementalGeneratorNode<T>
```
Comment (cston) [inline src/Compilers/Core/Portable/SourceGeneration/Nodes/IIncrementalGeneratorNode.cs:13]: >IIncrementalGeneratorNode [](http://example.com/codeflow?start=23&length=25)

Where is this implemented?

File: src/Compilers/Core/Portable/SourceGeneration/Nodes/DriverStateTable.cs [hunk 22]
```diff
@@ -0,0 +1,73 @@
...
+                {
+                    return (NodeStateTable<T>)_tableBuilder[source];
+                }
+
+                // get the previous table, if there was one for this node
+                NodeStateTable<T> previousTable = _previousTable._tables.ContainsKey(source)
+                                                  ? (NodeStateTable<T>)_previousTable._tables[source]
+                                                  : NodeStateTable<T>.Empty;
+
+                // request the node update its state based and store the new result
```
Comment (cston) [inline src/Compilers/Core/Portable/SourceGeneration/Nodes/DriverStateTable.cs:54]: >state based [](http://example.com/codeflow?start=47&length=11)

Should this be "state based on ..."?

File: src/Compilers/Core/Portable/SourceGeneration/Nodes/DriverStateTable.cs [hunk 23]
```diff
@@ -0,0 +1,73 @@
...
+                var newTable = source.UpdateStateTable(this, previousTable);
+                _tableBuilder[source] = newTable;
+                return newTable;
+            }
+
+            public DriverStateTable ToImmutable()
+            {
+                // we can compact the tables at this point, as we'll no longer be using them to determine current state
+                var keys = _tableBuilder.Keys.ToArray();
+                for (int i = 0; i < _tableBuilder.Count; i++)
```
Comment (cston) [inline src/Compilers/Core/Portable/SourceGeneration/Nodes/DriverStateTable.cs:64]: >for (int i = 0; i < _tableBuilder.Count; i++) [](http://example.com/codeflow?start=16&length=45)

`foreach (var key in keys)`

File: src/Compilers/Core/Portable/SourceGeneration/Nodes/IIncrementalGeneratorNode.cs [hunk 24]
```diff
@@ -0,0 +1,19 @@
...
+    {
+        NodeStateTable<T> UpdateStateTable(DriverStateTable.Builder graphState, NodeStateTable<T> previousTable);
+
+        IIncrementalGeneratorNode<T> WithComparer(IEqualityComparer<T> comparer);
```
Comment (cston) [inline src/Compilers/Core/Portable/SourceGeneration/Nodes/IIncrementalGeneratorNode.cs:17]: >WithComparer [](http://example.com/codeflow?start=37&length=12)

Where is `WithComparer()` used?

File: src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs [hunk 25]
```diff
@@ -0,0 +1,133 @@
...
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Microsoft.CodeAnalysis
+{
+    // NOTE:
+    /* A node table is the fundamental structure we use to track changes through the incremental 
+    generator api. It can be thought of as a series of slots that take their input from an
+    upstate table and produce 0-or-more outputs. When viewed from a downstream table the outputs
```
Comment (cston) [inline src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs:17]: >upstate [](http://example.com/codeflow?start=4&length=7)

"upstream"?

File: src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs [hunk 26]
```diff
@@ -0,0 +1,133 @@
...
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Microsoft.CodeAnalysis
+{
+    // NOTE:
```
Comment (cston) [inline src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs:14]: >// NOTE: [](http://example.com/codeflow?start=4&length=8)

Consider removing "NOTE" since it seems unnecessary, and consider using `// ` for the comment block.

File: src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs [hunk 27]
```diff
@@ -0,0 +1,133 @@
...
+
+    internal interface IStateTable
+    {
+        IStateTable Compact();
+    }
+
+    /// <summary>
+    /// A data structure that tracks the inputs and output of an execution node
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
```
Comment (cston) [inline src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs:55]: >T [](http://example.com/codeflow?start=25&length=1)

What does `T` represent?

File: src/Compilers/Core/Portable/SourceGeneration/Nodes/DriverStateTable.cs [hunk 28]
```diff
@@ -0,0 +1,73 @@
...
+                                                  ? (NodeStateTable<T>)_previousTable._tables[source]
+                                                  : NodeStateTable<T>.Empty;
+
+                // request the node update its state based and store the new result
+                var newTable = source.UpdateStateTable(this, previousTable);
+                _tableBuilder[source] = newTable;
+                return newTable;
+            }
+
+            public DriverStateTable ToImmutable()
```
Comment (cston) [inline src/Compilers/Core/Portable/SourceGeneration/Nodes/DriverStateTable.cs:60]: >ToImmutable [](http://example.com/codeflow?start=36&length=11)

Where is `ToImmutable()` called?

File: src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs [hunk 29]
```diff
@@ -0,0 +1,133 @@
...
+                // we only remove all entries at once, so only need to check the first item
+                if (entry.Length == 0 || entry[0].state != EntryState.Removed)
+                {
+                    compacted.Add(entry.SelectAsArray(e => (e.item, EntryState.Cached)));
+                }
+            }
+            return new NodeStateTable<T>(compacted.ToImmutableAndFree(), isCompacted: true, _exception);
+        }
+
+        public static NodeStateTable<T> FromFaultedTable<U>(NodeStateTable<U> table)
```
Comment (cston) [inline src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs:101]: >FromFaultedTable<U> [](http://example.com/codeflow?start=40&length=19)

This method doesn't appear to be used. Consider removing or adding a PROTOTYPE comment.

File: src/Compilers/Core/Portable/SourceGeneration/Nodes/ValueSourceExtensions.cs [hunk 30]
```diff
@@ -0,0 +1,43 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Microsoft.CodeAnalysis.SourceGeneration.Nodes
```
Comment (cston) [inline src/Compilers/Core/Portable/SourceGeneration/Nodes/ValueSourceExtensions.cs:9]: >.SourceGeneration.Nodes [](http://example.com/codeflow?start=32&length=23)

Is this correct?

File: src/Compilers/CSharp/Test/Semantic/SourceGeneration/GeneratorDriverTests.cs [hunk 31]
```diff
@@ -1120,5 +1120,59 @@ class C { }
...
+            catch (UserFunctionException e)
+            {
+                Assert.IsType<InvalidOperationException>(e.InnerException);
+            }
+
+            // cancellation is not wrapped, and is bubbled up
+            Assert.Throws<OperationCanceledException>(() => timeoutFunc(30));
+            Assert.Throws<OperationCanceledException>(() => userTimeoutFunc(30));
+
+            Func<int, int> makeFunc(bool throwEx = false, bool throwCanceled = false)
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/SourceGeneration/GeneratorDriverTests.cs:1158]: >makeFunc [](http://example.com/codeflow?start=27&length=8)

It looks like this local function has three different cases below, and just three distinct callers. Would it be simpler just to inline the particular cases in the each caller?

File: src/Compilers/CSharp/Test/Semantic/SourceGeneration/StateTableTests.cs [hunk 32]
```diff
@@ -0,0 +1,260 @@
...
+            var builder = new NodeStateTable<int>.Builder();
+            builder.AddEntries(ImmutableArray.Create(1, 2, 3), EntryState.Added);
+            builder.AddEntries(ImmutableArray.Create(4, 5, 6), EntryState.Removed);
+            builder.AddEntries(ImmutableArray.Create(7, 8, 9), EntryState.Added);
+            var table = builder.ToImmutableAndFree();
+
+            var expected = ImmutableArray.Create((1, EntryState.Added), (2, EntryState.Added), (3, EntryState.Added), (4, EntryState.Removed), (5, EntryState.Removed), (6, EntryState.Removed), (7, EntryState.Added), (8, EntryState.Added), (9, EntryState.Added));
+            AssertTableEntries(table, expected);
+
+            var compactedTable = (NodeStateTable<int>)table.Compact();
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/SourceGeneration/StateTableTests.cs:88]: >(NodeStateTable<int>)table.Compact [](http://example.com/codeflow?start=33&length=34)

Should `Compact()` return `NodeStateTable<T>`?

File: src/Compilers/CSharp/Test/Semantic/SourceGeneration/StateTableTests.cs [hunk 33]
```diff
@@ -0,0 +1,260 @@
...
+            DriverStateTable.Builder builder = new DriverStateTable.Builder(DriverStateTable.Empty);
+            builder.GetLatestStateTableForNode(callbackNode);
+
+            Assert.Same(NodeStateTable<int>.Empty, passedIn);
+
+            // gives the returned table the second time around
+            DriverStateTable.Builder builder2 = new DriverStateTable.Builder(builder.ToImmutable());
+            builder2.GetLatestStateTableForNode(callbackNode);
+
+            Debug.Assert(passedIn is object);
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/SourceGeneration/StateTableTests.cs:170]: >Debug.Assert [](http://example.com/codeflow?start=12&length=12)

`Assert.NotNull()`

File: src/Compilers/CSharp/Test/Semantic/SourceGeneration/StateTableTests.cs [hunk 34]
```diff
@@ -0,0 +1,260 @@
...
+            DriverStateTable.Builder builder = new DriverStateTable.Builder(DriverStateTable.Empty);
+            builder.GetLatestStateTableForNode(callbackNode);
+            Assert.Same(NodeStateTable<int>.Empty, passedIn);
+
+            // gives the returned table the second time around
+            DriverStateTable.Builder builder2 = new DriverStateTable.Builder(builder.ToImmutable());
+            builder2.GetLatestStateTableForNode(callbackNode);
+
+            // table returned from the first instance was compacted by the builder
+            Debug.Assert(passedIn is object);
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/SourceGeneration/StateTableTests.cs:201]: >Debug.Assert [](http://example.com/codeflow?start=12&length=12)

`Assert.NotNull()`

File: src/Compilers/CSharp/Test/Semantic/SourceGeneration/StateTableTests.cs (hunk 32 above)
Comment (chsienki) [reply src/Compilers/CSharp/Test/Semantic/SourceGeneration/StateTableTests.cs:88]: Yeah, this is tricky, because `Compact()` is called from the `DriverStateTable`, which doesn't know the `T`. I ended up making a non-generic interface with `Compact()` that returns that interface so it could call it in a type agnostic manner; all the other use cases we know the `T` so can cast it back. 

Open to suggestions on better approaches though.

File: src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs [hunk 35]
```diff
@@ -0,0 +1,132 @@
...
+                if (entry.Length == 0 || entry[0].state != EntryState.Removed)
+                {
+                    compacted.Add(entry.SelectAsArray(e => (e.item, EntryState.Cached)));
+                }
+            }
+            return new NodeStateTable<T>(compacted.ToImmutableAndFree(), isCompacted: true, _exception);
+        }
+
+        // PROTOTYPE: this will be called to allow exceptions to flow through the graph
+        public static NodeStateTable<T> FromFaultedTable<U>(NodeStateTable<U> table)
```
Comment (cston) [inline src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs:100]: >U [](http://example.com/codeflow?start=75&length=1)

Should this be `T`? (When would `T` and `U` be distinct?)

File: src/Compilers/CSharp/Test/Semantic/SourceGeneration/StateTableTests.cs (hunk 32 above)
Comment (cston) [reply src/Compilers/CSharp/Test/Semantic/SourceGeneration/StateTableTests.cs:86]: Perhaps `NodeStateTable<T>` could implement the interface explicitly.
```csharp
public NodeStateTable<T> Compact()
{
    // ...
}

IStateTable IStateTable.Compact()
{
    return Compact();
}
```

File: src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs (hunk 35 above)
Comment (chsienki) [reply src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs:100]: When a 'downstream' `table<T>` is trying to get its values from an 'upstream' `table<U>`. `table<U>` is faulted, so we essentially just say, make a `table<T>` with the same fault.

**Review by cston — APPROVED** [review]

File: src/Compilers/CSharp/Test/Semantic/SourceGeneration/StateTableTests.cs (hunk 32 above)
Comment (chsienki) [reply src/Compilers/CSharp/Test/Semantic/SourceGeneration/StateTableTests.cs:86]: Oh, yeah thats neat. thanks

**Comment by RikkiGibson** [comment]
It looks like there are PublicAPI analyzer failures.

File: src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs (hunk 13 above)
Comment (chsienki) [reply src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs:92]: Added a proto comment to track this

File: src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs (hunk 9 above)
Comment (chsienki) [reply src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs:125]: Added a proto comment to track this. It's changed slightly in a future PR anyway, but there are some good cases to consider + test, so want to make sure we don't lose track of them.

File: src/Compilers/Core/Portable/SourceGeneration/GeneratorAdaptor.cs [hunk 36]
```diff
@@ -0,0 +1,45 @@
...
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Microsoft.CodeAnalysis.SourceGeneration
+{
+    /// <summary>
+    /// Adapts an ISourceGenerator to an incremental generator that
+    /// by providng an execution environment that matches the old one
```
Comment (RikkiGibson) [inline src/Compilers/Core/Portable/SourceGeneration/GeneratorAdaptor.cs:13]: I am the typo master..this should be "providing".

File: src/Compilers/Core/Portable/SourceGeneration/IncrementalValueSource.cs [hunk 37]
```diff
@@ -0,0 +1,29 @@
...
+using System.Text;
+
+namespace Microsoft.CodeAnalysis
+{
+    /// <summary>
+    /// Represents a value that can have transforms applied to it to construct an execution pipeline
+    /// </summary>
+    /// <remarks>
+    /// This is an opaque type that cannot be used directly. Instead an <see cref="IIncrementalGenerator" />
+    /// will receive a set of value sources when constructing it's execution pipeline. A set of extension methods
```
Comment (RikkiGibson) [inline src/Compilers/Core/Portable/SourceGeneration/IncrementalValueSource.cs:16]: "it's"->"its"

**Review by RikkiGibson — APPROVED** [review]

File: src/Compilers/CSharp/Test/Semantic/SourceGeneration/StateTableTests.cs [hunk 38]
```diff
@@ -0,0 +1,258 @@
...
+            builder.AddEntries(ImmutableArray.Create(2, 3), EntryState.Cached);
+            builder.AddEntries(ImmutableArray.Create(4, 5), EntryState.Modified);
+            builder.AddEntries(ImmutableArray.Create(6), EntryState.Added);
+            var previousTable = builder.ToImmutableAndFree();
+
+            builder = new NodeStateTable<int>.Builder();
+            builder.AddEntries(ImmutableArray.Create(10, 11), EntryState.Added);
+            builder.AddEntriesFromPreviousTable(previousTable, EntryState.Cached); // ((2, EntryState.Cached), (3, EntryState.Cached))
+            builder.AddEntries(ImmutableArray.Create(20, 21, 22), EntryState.Modified);
+            builder.AddEntriesFromPreviousTable(previousTable, EntryState.Removed); //((6, EntryState.Removed)));
```
[the remark this answers is not in the document]
Comment (chsienki) [reply src/Compilers/CSharp/Test/Semantic/SourceGeneration/StateTableTests.cs:67]: There are four entries in the original table: 

slot | values 
-|-
1 | (1, added)
2 | (2, cached), (3, cached)
3 | (4, modified), (5, modified)
4 | (6, Added)

The second table puts in its own entries for slots 1 and 3, and copies the entries from slots 2 and 4. When it copies the entries it has a chance to override the state of the previous entries. for slot 2 it overrides with cached, which happens to be the same. For slot 4 for it overrides with removed.

[the remark this answers is not in the document]
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/SourceGeneration/StateTableTests.cs:67]: Thanks, this clears up a lot about how "slots" are assigned/tracked :)
